<a href="https://colab.research.google.com/github/DanielNevesPP/IntroducaoAComputacaoNumerica/blob/main/Int_Comp_Num%C3%A9rica_Trabalho_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Trabalho prático # 4
##Introdução à Computação Numérica

###Tópico: Integração numérica 

---

Nome: Daniel Neves Pereira Pinto

DRE: 119061593


---

**Instruções:** 

**Preencha o notebook** com seus dados, seus códigos e **rode todos os códigos** 

**Inclua  um relatório, na parte final do notebook,** com suas conclusões, respostas, observações e comentários relacionados com os tópicos e questões abordadas nas tarefas.  

**A elaboração dos códigos pode ser realizada em duplas, mas o resto do trabalho é individual.**

*Se você fez o seu trabalho em dupla inclua essa informação no relatório. Além disso, se você  também recebeu ajuda de um(a) colega/amigo(a) ou consultou/usou informações da internet ou outras fontes, coloque referências à elas no seu relatório.*


**Entrega do trabalho:**

*Data de entrega: 8/dezembro/2022* 

---


## Tarefa #1

Implemente funções em Python para aproximar a integral definida
$$
I = \int_a^b f(x)\,dx,
$$
usando: 
i) a regra dos trapézios repetida/composta;
ii) a regra de Simpson repetida/composta.

a) Considere como parâmetros de entrada o espaçamento $h$ entre os nós (pontos) que serão usados e uma lista com os valores da função nesses nós: $f_j = f(a+j\, h)$,  $j=0,\dots,n$.

b) Considere como parâmetros de entrada o intervalo $[a, b]$, a função $f$ e a quantidade de nós (pontos) que serão usados.

**Observação**: Você precisará implementar 4 funções.


In [ ]:
#a)
def trapezios1(h, lista):
  #inicializa a variavel soma com 0
  soma = 0
  #faz a soma dos nos da lista excluindo o primeiro e o ultimo
  for i in range(1, len(lista)-1):
    soma += lista[i]
  #retorna o calculo da regra do trapezio repetida/composta
  return (h/2) * (lista[0] + lista[-1] + 2*soma)



def simpson1(h, lista):
  #inicializa a variavel soma1 com 0
  soma1 = 0
  #faz a soma dos nos da lista de 2 em 2, comecando no no da posicao 2 e
  #terminando no da antepenultima posicao
  for i in range(1, int((len(lista)-1)/2)):
    soma1 += lista[2*i]

  #inicializa a variavel soma2 com 0
  soma2 = 0
  #faz a soma dos nos da lista de 2 em 2, comecando no no da posicao 1 e
  #terminando no da penultima posicao
  for i in range(1, int((len(lista)-1)/2)+1):
    soma2+= lista[(2*i)-1]

  #retorna o calculo da regra de simpson repetida/composta
  #lista[0] eh o primeiro item da lista e lista[-1] o ultimo
  return (h/3) * (lista[0] + lista[-1] + 2*soma1 + 4*soma2)

################################################################################

#b)
def trapezios2(a, b, f, n):
  #calcula h baseado em a, b e n
  h = (b-a)/n

  #inicializa a variavel soma com 0
  soma = 0
  #faz a soma dos nos, excluindo o primeiro e o ultimo
  for i in range(1, n):
    soma += f(a + i*h)

  #retorna o calculo da regra dos trapezios repetida/composta
  return (h/2) * (f(a) + f(b) + 2*soma)



def simpson2(a, b, f, n):
  #calcula h baseado em a, b e n
  h = (b-a)/n

  #inicializa a variavel soma1 com 0
  soma1 = 0
  #faz a soma dos nos de 2 em 2, comecando em f(x2) e terminando em f(x(n-2))
  for i in range(1, int((n/2))):
    soma1 += f(a + 2*i*h)

  #inicializa a variavel soma2 com 0
  soma2 = 0
  #faz a soma dos nos de 2 em 2, comecando em f(x1) e terminando em f(x(n-1))
  for i in range(1, int(n/2)+1):
    soma2 += f(a + (2*i - 1)*h)

  #retorna o calculo da regra de simpson repetida/composta
  return (h/3) * (f(a) + f(b) + 2*soma1 + 4*soma2)

##Tarefa #2: Teste as funções implementadas

a) Procure uma função $f$ cuja integral definida pode ser calculada analiticamente (de forma exata) no intervalo escolhido.

b) Obtenha aproximações do valor da integral definida dessa função considerando $n=10$, $50$, $100$ e $200$ (note que a quantidade de nós em cada caso será igual a $n+1$). Para cada valor de $n$ obtenha aproximações usando suas implementações das regras dos trapézios e de Simpson.  

c) Determine os erros absolutos associados com as aproximações obtidas no item a). 

d) Discuta e explique os resultados obtidos em b) e c).

In [ ]:
#a)
#valor de n
n = 20000

#funcao f escolhida:
f = lambda x: 4*x**2

#intervalo escolhido: [1, 2]
a = 1
b = 2
#calculando h baseado em a,b e n
h = (b-a)/n

#criando a lista de nos:
lista = []
for i in range(n+1):
  lista.append(f(a + i*h))

#valor exato da integral de f nesse intervalo: 28/3 = 9,333...

################################################################################

#b)
print("Regra do trapezio forma 1:", trapezios1(h, lista))
print("Regra do trapezio forma 2:", trapezios2(a, b, f, n))
print("Regra de Simpson forma 1:", simpson1(h, lista))
print("Regra de Simpson forma 2:", simpson2(a, b, f, n))

################################################################################

#c)
print()
print("28/3 =", 28/3)
print("Erro absoluto da regra do trapezio na forma 1:", abs((28/3) - trapezios1(h, lista)))
print("Erro absoluto da regra do trapezio na forma 2:", abs((28/3) - trapezios2(a, b, f, n)))
print("Erro absoluto da regra de Simpson na forma 1:", abs((28/3) - simpson1(h, lista)))
print("Erro absoluto da regra de Simpson na forma 2:", abs((28/3) - simpson2(a, b, f, n)))

Regra do trapezio forma 1: 9.333333335000008
Regra do trapezio forma 2: 9.333333335000008
Regra de Simpson forma 1: 9.3333333333333
Regra de Simpson forma 2: 9.3333333333333

28/3 = 9.333333333333334
Erro absoluto da regra do trapezio na forma 1: 1.6666739099946426e-09
Erro absoluto da regra do trapezio na forma 2: 1.6666739099946426e-09
Erro absoluto da regra de Simpson na forma 1: 3.375077994860476e-14
Erro absoluto da regra de Simpson na forma 2: 3.375077994860476e-14


##Tarefa #3: Aplicação

A **função erro** é definida como
$$
\mathrm{erf}(x) = \frac{2}{\sqrt{\pi}}\int_0^x \mathrm{e}^{-t^2}\,dt.
$$

a) Implemente uma função que determine o valor aproximado de $\mathrm{erf}(x)$ para qualquer entrada $x$ com um erro absoluto menor que $10^{-8}$. Explique as ideias básicas da sua implementação.

b) Teste a função implementada comparando com os seguintes valores*: 

i) $\mathrm{erf}(-2) = -0.9953222650189527$, 

ii) $\mathrm{erf}(0.5) = 0.5204998778130465$, 

iii) $\mathrm{erf}(1) = 0.8427007929497149$, 

iv) $\mathrm{erf}(5) = 0.9999999999984625$.

*Os valores apresentados foram calculados com 16 dígitos de precisão.




In [ ]:
#a)
import math

#funcao dentro da integral
f = lambda t: math.e**(-(t**2))
#valor de n usado
n = 100

#funcao erro feita com a regra de simpson
def erf(x):
  return (2/math.sqrt(math.pi))*simpson2(0, x, f, n)

################################################################################

#b)
print("erf(-2):", erf(-2))
print("erf(0.5):", erf(0.5))
print("erf(1):", erf(1))
print("erf(5):", erf(5))
print()
print("Erro absoluto de erf(-2):", abs(-0.9953222650189527 - erf(-2)))
print("Erro absoluto de erf(0.5):", abs(0.5204998778130465 - erf(0.5)))
print("Erro absoluto de erf(1):", abs(0.8427007929497149 - erf(1)))
print("Erro absoluto de erf(5):", abs(0.9999999999984625 - erf(5)))

erf(-2): -0.9953222642841127
erf(0.5): 0.5204998778283034
erf(1): 0.8427007930419588
erf(5): 0.9999999999984617

Erro absoluto de erf(-2): 7.348399666540217e-10
Erro absoluto de erf(0.5): 1.5256906849003826e-11
Erro absoluto de erf(1): 9.22438792017033e-11
Erro absoluto de erf(5): 7.771561172376096e-16


## Relatório

# **Tarefa #1:**
O código está feito e comentado

# **Tarefa #2:**
### **a)**
A função escholhida foi f(x) = 4*(x^2) no intervalo [1, 2], cuja integral definida nesse intervalo é 28/3 = 9.333333333333334 (calculado pelo python)

### **b)**
n = 10:

Regra do trapézio (ambas as formas): 9.340000000000002

Regra de Simpson (ambas as formas): 9.333333333333334

n = 50:

Regra do trapézio (ambas as formas): 9.3336

Regra de Simpson (ambas as formas): 9.333333333333334

n = 100:

Regra do trapézio (ambas as formas): 9.333400000000005

Regra de Simpson (ambas as formas): 9.333333333333334

n = 200:

Regra do trapezio (ambas as formas): 9.333349999999998

Regra de Simpson (ambas as formas): 9.333333333333337

### **c)**
n = 10:

Erro absoluto da regra do trapezio (ambas as formas): 0.006666666666667709

Erro absoluto da regra de Simpson (ambas as formas): 0.0

n = 50:

Erro absoluto da regra do trapezio (ambas as formas): 0.0002666666666666373

Erro absoluto da regra de Simpson (ambas as formas): 0.0

n = 100:

Erro absoluto da regra do trapezio (ambas as formas): 6.666666667065613e-05

Erro absoluto da regra de Simpson (ambas as formas): 0.0

n = 200:

Erro absoluto da regra do trapezio (ambas as formas): 1.666666666366723e-05

Erro absoluto da regra de Simpson (ambas as formas): 3.552713678800501e-15

### **d)**
Pelos resultados obtidos, podemos ver que a regra de Simpson atingiu um valor mais próximo em todos os valores de n usados, atingindo o valor exato em n = 10, 50 e 100. Já a regra dos trapézios foi atingindo valores cada vez mais próximos do valor exato conforme n foi aumentando. Portanto, pelo menos para essa função usada, a regra de Simpson é mais efetiva (principalmente quando n não é muito grande), enquanto a regra dos trapézios, apesar de ser menos exata, atinge valores muito próximos com valores de n maiores.

# **Tarefa #3:**
### **a)**
Para fazer a função erro, resolvi usar a regra de Simpson, pois foi a que obteve melhores resultados na tarefa #2, assim como o n = 100.

### **b)**
Valor encontrado em erf(-2): -0.9953222642841127

Valor encontrado em erf(0.5): 0.5204998778283034

Valor encontrado em erf(1): 0.8427007930419588

Valor encontrado em erf(5): 0.9999999999984617

Erro absoluto de erf(-2): 7.348399666540217e-10

Erro absoluto de erf(0.5): 1.5256906849003826e-11

Erro absoluto de erf(1): 9.22438792017033e-11

Erro absoluto de erf(5): 7.771561172376096e-16

Todos os erros encontrados foram menores que 10^(-8)
